ArmPkg/Drivers/SmmuDxe/SmmuDxe.c

```diff
@@ -0,0 +1,774 @@
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to set global bypass.\n", __func__));
+  }
+}
+
+/**
+  Initialize the SMMU_INFO structure.
+
+  @retval Pointer to the allocated SMMU_INFO structure.
+**/
+SMMU_INFO *
+EFIAPI
+SmmuInit (
+  VOID
+  )
+{
+  return (SMMU_INFO *)AllocateZeroPool (sizeof (SMMU_INFO));
+}
+
+/**
+  Deinitialize and free the SMMU_INFO structure and everything inside.
+
+  @param [in]  Smmu    Pointer to the SMMU_INFO structure to deinitialize.
+**/
+VOID
+EFIAPI
+SmmuDeInit (
+  SMMU_INFO  *SmmuInfo
+  )
+{
+  EFI_STATUS  Status;
+
+  Status = SmmuV3DisableTranslation (mSmmu->SmmuBase);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to disable SMMUv3 translation\n"));
+  }
+
+  Status = SmmuV3GlobalAbort (mSmmu->SmmuBase);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "Failed to global abort SMMUv3\n"));
+  }
+
+  PageTableDeInit (0, SmmuInfo->PageTableRoot);
+  SmmuInfo->PageTableRoot = NULL;
+  SmmuV3FreeStreamTable ((SMMUV3_STREAM_TABLE_ENTRY *)SmmuInfo->StreamTable, SmmuInfo->StreamTableSize);
+  SmmuInfo->StreamTable = NULL;
+  SmmuV3FreeQueue (SmmuInfo->CommandQueue);
+  SmmuInfo->CommandQueue = NULL;
+  SmmuV3FreeQueue (SmmuInfo->EventQueue);
+  SmmuInfo->EventQueue = NULL;
+  FreePool (SmmuInfo);
+}
+
+/**
+  Entrypoint for SmmuDxe driver.
+
+  @param [in] ImageHandle    The firmware allocated handle for the EFI image.
+  @param [in] SystemTable    A pointer to the EFI System Table.
+
+  @retval EFI_SUCCESS        The entry point is executed successfully.
+  @retval Others             Some error occurs when executing this entry point.
+**/
+EFI_STATUS
+EFIAPI
+InitializeSmmuDxe (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS               Status;
+  EFI_EVENT                Event;
+  EFI_ACPI_TABLE_PROTOCOL  *AcpiTable;
+  SMMU_CONFIG              *SmmuConfig;
+
+  SmmuConfig = GetSmmuConfigHobData ();
+
+  if (SmmuConfig == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to get SMMU config data from gSmmuConfigGuid\n", __func__));
+    return EFI_NOT_FOUND;
+  }
+
+  // Check if ACPI Table Protocol has been installed
+  Status = gBS->LocateProtocol (
+                  &gEfiAcpiTableProtocolGuid,
+                  NULL,
+                  (VOID **)&AcpiTable
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to locate ACPI Table Protocol\n",
+      __func__
+      ));
+    return Status;
+  }
+
+  Status = gBS->CreateEventEx (
+                  EVT_NOTIFY_SIGNAL,
+                  TPL_NOTIFY,
+                  SmmuV3ExitBootServices,
+                  NULL,
+                  &gEfiEventExitBootServicesGuid,
+                  &Event
+                  );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to create ExitBootServices event\n", __func__));
+    return Status;
+  }
+
+  mSmmu = SmmuInit ();

```

Comment:
nit: this just wraps an allocate call, can just do that here

Reply:
SmmuInit now takes in the SmmuBase as a input, so keeping this function.